demos/mqtt/mqtt_demo_basic_tls/mqtt_demo_basic_tls.c

```diff
@@ -253,83 +630,421 @@ static SSL * tlsSetup( int tcpSocket )
+{
+    int status = EXIT_SUCCESS;
+    MQTTStateCursor_t cursor = 0U;
+    MQTTStatus_t mqttStatus = MQTTSuccess;
+    uint16_t packetId = MQTT_PACKET_ID_INVALID;
+    uint8_t index = MAX_OUTGOING_PUBLISHES;
+
+    assert( outgoingPublishPackets != NULL );
+
+    /* Check all the QoS2 publish packets that are still waiting for
+     * a PUBREC. This can be identified by querying the state machine
+     * for all the outgoing publish packets which are in state
+     * #MQTTPubRecPending. */
+    do
+    {
+        /* Find the packet ids which are in #MQTTPubRecPending state.
+         * Using the same cursor will help restart the search from the
+         * last searched position. */
+        packetId = MQTT_StateSelect( pContext, MQTTPubRecPending, &cursor );
+
+        /* Check if a valid packet id is obtained. */
+        if( packetId != MQTT_PACKET_ID_INVALID )
+        {
+            /* Resending the publish after marking publish as duplicate. */
+            index = getIndexOfOutgoingPublishWithPacketId( packetId );
+
+            /* Check if a valid index is obtained. */
+            if( index < MAX_OUTGOING_PUBLISHES )
+            {
+                /*Resend the publish after marking it as duplicate. */
+                outgoingPublishPackets[ index ].pubInfo.dup = true;
+
+                LogInfo( ( "Sending duplicate PUBLISH with packet id %u.",
+                           packetId ) );
+                mqttStatus = MQTT_Publish( pContext,
+                                           &outgoingPublishPackets[ index ].pubInfo,
+                                           packetId );
+
+                if( mqttStatus != MQTTSuccess )
+                {
+                    LogError( ( "Sending duplicate PUBLISH for packet id %u "
+                                " failed with status %u.",
+                                packetId,
+                                mqttStatus ) );
+                    status = EXIT_FAILURE;
+                    break;
+                }
+                else
+                {
+                    LogInfo( ( "Sent duplicate PUBLISH successfully for packet id %u.\n\n",
+                               packetId ) );
+                }
+            }
+            else
+            {
+                LogError( ( "Failed to obtain a stored outgoing publish packet for"
+                            " packet id %u.",
+                            packetId ) );
+                status = EXIT_FAILURE;
+            }
+        }
+    } while( packetId != MQTT_PACKET_ID_INVALID );
+
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+static void handleIncomingPublish( MQTTPublishInfo_t * pPublishInfo,
+                                   uint16_t packetIdentifier )
+{
+    assert( pPublishInfo != NULL );
+
+    /* Process incoming Publish. */
+    LogInfo( ( "Incoming QOS : %d.", pPublishInfo->qos ) );
+
+    /* Verify the received publish is for the we have subscribed to. */
```

Comment:
Since the demo is  receiving echo of Publish message sent, may be we should verify packetId as well.  Store lastPacketIdSent and compare it with received packet id.

Reply:
Those packet Ids won't be the same. That will be part of 2 independent publishes. Spec only mentions about maintaining the packet Ids for Publishes and the acks for it.

Reply:
If you have verified it then I'm OK. Thanks.